sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PartitionPruning.scala

```diff
@@ -0,0 +1,264 @@
+      // filter out expressions with more than one attribute on any side of the operator
+      case (leftAttr :: Nil, rightAttr :: Nil)
+        if SQLConf.get.dynamicPartitionPruningWithStats =>
+          // get the CBO stats for each attribute in the join condition
+          val partDistinctCount = distinctCounts(leftAttr, partPlan)
+          val otherDistinctCount = distinctCounts(rightAttr, otherPlan)
+          val availableStats = partDistinctCount.isDefined && partDistinctCount.get > 0 &&
+            otherDistinctCount.isDefined
+          if (!availableStats) {
+            fallbackRatio
+          } else if (partDistinctCount.get.toDouble <= otherDistinctCount.get.toDouble) {
+            // there is likely an estimation error, so we fallback
+            fallbackRatio
+          } else {
+            1 - otherDistinctCount.get.toDouble / partDistinctCount.get.toDouble
+          }
+      case _ => fallbackRatio
+    }
+
+    // the pruning overhead is the total size in bytes of all scan relations
+    val overhead = otherPlan.collectLeaves().map(_.stats.sizeInBytes).sum.toFloat
+
+    filterRatio * partPlan.stats.sizeInBytes.toFloat > overhead.toFloat
+  }
+
+  /**
+   * Returns whether an expression is likely to be selective
+   */
+  private def isLikelySelective(e: Expression): Boolean = e match {
+    case Not(expr) => isLikelySelective(expr)
+    case And(l, r) => isLikelySelective(l) || isLikelySelective(r)
+    case Or(l, r) => isLikelySelective(l) && isLikelySelective(r)
+    case Like(_, _) => true
+    case _: BinaryComparison => true
+    case _: In | _: InSet => true
+    case _: StringPredicate => true
+    case _ => false
+  }
+
+  /**
+   * Search a filtering predicate in a given logical plan
+   */
+  private def hasSelectivePredicate(plan: LogicalPlan): Boolean = {
+    plan.find {
+      case f: Filter => isLikelySelective(f.condition)
+      case _ => false
+    }.isDefined
+  }
+
+  /**
+   * To be able to prune partitions on a join key, the filtering side needs to
+   * meet the following requirements:
+   *   (1) it can not be a stream
+   *   (2) it needs to contain a selective predicate used for filtering
+   *   (3) the join key is in the attribute set of the filtering plan
+   */
+  private def hasPartitionPruningFilter(key: Expression, plan: LogicalPlan): Boolean = {
+    !plan.isStreaming && hasSelectivePredicate(plan) &&
+      key.references.subsetOf(plan.outputSet)
+  }
+
+  private def canPruneLeft(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | RightOuter => true
+    case _ => false
+  }
+
+  private def canPruneRight(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftOuter => true
+    case _ => false
+  }
+
+  private def prune(plan: LogicalPlan): LogicalPlan = {
+    plan transformUp {
+      // skip this rule if there's already a DPP subquery on the LHS of a join
```

Comment:
Does this really help? once filter pushdown is applied, it's hard to make this rule idempotent.